datafusion/physical-plan/src/sorts/batches.rs

```diff
@@ -0,0 +1,31 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use super::cursor::{Cursor, CursorValues};
+
+#[derive(Debug)]
+pub(crate) struct BatchCursor<C: CursorValues> {
+    /// The index into SortOrderBuilder::batches
+    /// TODO: this will become a BatchId, for record batch collected (and not passed across streams)
+    pub batch_idx: usize,
```

Comment:
Note: the BatchId is required for the batch tracking. Therefore, we have to yield something (beyond just the CursorValues) which contains this tracking id.

I decided to have the same abstraction (a.k.a. BatchCursor) be used for both batches currently being sorted (in a merge/loser tree), and the batches being yielded to the next merge/loser tree node (in the cascade tree). But we could revisit this design.